homeassistant/components/here_travel_time/sensor.py

```diff
@@ -0,0 +1,468 @@
+            return ICON_PEDESTRIAN
+        if self._here_data.travel_mode in TRAVEL_MODES_PUBLIC:
+            return ICON_PUBLIC
+        if self._here_data.travel_mode == TRAVEL_MODE_TRUCK:
+            return ICON_TRUCK
+        return ICON_CAR
+
+    async def async_update(self) -> None:
+        """Update Sensor Information."""
+        # Convert device_trackers to HERE friendly location
+        if self._origin_entity_id is not None:
+            self._here_data.origin = await self._get_location_from_entity(
+                self._origin_entity_id
+            )
+
+        if self._destination_entity_id is not None:
+            self._here_data.destination = await self._get_location_from_entity(
+                self._destination_entity_id
+            )
+
+        await self._hass.async_add_executor_job(self._here_data.update)
+
+    async def _get_location_from_entity(self, entity_id: str) -> Optional[str]:
+        """Get the location from the entity state or attributes."""
+        entity = self._hass.states.get(entity_id)
+
+        if entity is None:
+            _LOGGER.error("Unable to find entity %s", entity_id)
+            return None
+
+        # Check if the entity has location attributes
+        if location.has_location(entity):
+            return self._get_location_from_attributes(entity)
+
+        # Check if device is in a zone
+        zone_entity = self._hass.states.get("zone.{}".format(entity.state))
+        if location.has_location(zone_entity):
+            _LOGGER.debug(
+                "%s is in %s, getting zone location", entity_id, zone_entity.entity_id
+            )
+            return self._get_location_from_attributes(zone_entity)
+
+        # If zone was not found in state then use the state as the location
+        if entity_id.startswith("sensor."):
+            return entity.state
+
+    @staticmethod
+    def _get_location_from_attributes(entity: State) -> str:
+        """Get the lat/long string from an entities attributes."""
+        attr = entity.attributes
+        return "{},{}".format(attr.get(ATTR_LATITUDE), attr.get(ATTR_LONGITUDE))
+
+
+class HERETravelTimeData:
+    """HERETravelTime data object."""
+
+    def __init__(
+        self,
+        origin: None,
+        destination: None,
+        app_id: str,
+        app_code: str,
+        travel_mode: str,
+        traffic_mode: bool,
+        route_mode: str,
+        units: str,
+    ) -> None:
+        """Initialize herepy."""
+        self.origin = origin
+        self.destination = destination
+        self.travel_mode = travel_mode
+        self.traffic_mode = traffic_mode
+        self.route_mode = route_mode
+        self.attribution = None
+        self.traffic_time = None
+        self.distance = None
+        self.route = None
+        self.base_time = None
+        self.origin_name = None
+        self.destination_name = None
+        self.units = units
+        self._client = herepy.RoutingApi(app_id, app_code)
```

Comment:
Can the app id and code be incorrect? If they can we should check that early, log error and return in setup_platform. We also try to avoid side effects in init method. It's clearer if pass in the client instance to the data instance.

Reply:
They can be incorrect but this can only be checked when an actual request is made, not when the client is instantiated. I would have to make a dummy call against a hardcoded set of coordinates and check for the error code.

Reply:
It's fine to do a request.